stmt/stmts/insert.go

```diff
@@ -291,36 +262,96 @@ func (s *InsertIntoStmt) Exec(ctx context.Context) (_ rset.Recordset, err error)
 		// `t(id int AUTO_INCREMENT, c1 int, PRIMARY KEY (id))`
 		// `insert t (c1) values(1),(2),(3);`
 		// Last insert id will be 1, not 3.
-		h, err := t.AddRecord(ctx, r)
+		h, err := t.AddRecord(ctx, row)
 		if err == nil {
 			continue
 		}
+
 		if len(s.OnDuplicate) == 0 || !errors2.ErrorEqual(err, kv.ErrKeyExists) {
 			return nil, errors.Trace(err)
 		}
-		// On duplicate key Update the duplicate row.
-		// Evaluate the updated value.
-		// TODO: report rows affected and last insert id.
-		data, err := t.Row(ctx, h)
-		if err != nil {
+		if err = execOnDuplicateUpdate(ctx, t, row, h, toUpdateColumns); err != nil {
 			return nil, errors.Trace(err)
 		}
+	}
 
-		toUpdateArgs[expression.ExprEvalValuesFunc] = func(name string) (interface{}, error) {
-			c, err1 := findColumnByName(t, name)
-			if err1 != nil {
-				return nil, errors.Trace(err1)
-			}
-			return r[c.Offset], nil
+	return nil, nil
+}
+
+func (s *InsertIntoStmt) checkValueCount(insertValueCount, valueCount, num int, cols []*column.Col) error {
+	if insertValueCount != valueCount {
+		// "insert into t values (), ()" is valid.
+		// "insert into t values (), (1)" is not valid.
+		// "insert into t values (1), ()" is not valid.
+		// "insert into t values (1,2), (1)" is not valid.
+		// So the value count must be same for all insert list.
+		return errors.Errorf("Column count doesn't match value count at row %d", num+1)
+	}
+	if valueCount == 0 && len(s.ColNames) > 0 {
+		// "insert into t (c1) values ()" is not valid.
+		return errors.Errorf("INSERT INTO %s: expected %d value(s), have %d", s.TableIdent, len(s.ColNames), 0)
+	} else if valueCount > 0 && valueCount != len(cols) {
+		return errors.Errorf("INSERT INTO %s: expected %d value(s), have %d", s.TableIdent, len(cols), valueCount)
+	}
+
+	return nil
+}
+
+func (s *InsertIntoStmt) getRow(ctx context.Context, t table.Table, cols []*column.Col, list []expression.Expression, m map[interface{}]interface{}) ([]interface{}, error) {
+	r := make([]interface{}, len(t.Cols()))
+	marked := make(map[int]struct{}, len(list))
+	for i, expr := range list {
+		// For "insert into t values (default)" Default Eval.
+		m[expression.ExprEvalDefaultName] = cols[i].Name.O
+
+		val, evalErr := expr.Eval(ctx, m)
+		if evalErr != nil {
+			return nil, errors.Trace(evalErr)
 		}
+		r[cols[i].Offset] = val
+		marked[cols[i].Offset] = struct{}{}
+	}
 
-		err = updateRecord(ctx, h, data, t, toUpdateColumns, toUpdateArgs, 0, true)
-		if err != nil {
-			return nil, errors.Trace(err)
+	// Clear last insert id.
+	variable.GetSessionVars(ctx).SetLastInsertID(0)
+
+	err := s.initDefaultValues(ctx, t, r, marked)
```

Comment:
Why not use `if err := func(); err != nil` format? How about defining an error variable above?